cpp/src/lance/encodings/fixed_size_binary.cc

```diff
@@ -0,0 +1,124 @@
+    return plain_encoder.Write(list_arr->values());
+  }
+
+  return values_position;
+}
+
+std::string FixedSizeBinaryEncoder::ToString() const { return "FixedSizeBinaryEncoder"; }
+
+::arrow::Result<std::shared_ptr<::arrow::Array>> FixedSizedBinaryDecoder::ToFixedSizeBinaryArray(
+    int32_t start, int32_t length) const {
+  auto bytes = type_->byte_width();
+  ARROW_ASSIGN_OR_RAISE(auto buf, infile_->ReadAt(position_ + start * bytes, length * bytes));
+  return std::make_shared<::arrow::FixedSizeBinaryArray>(type_, length, buf);
+}
+
+::arrow::Result<std::shared_ptr<::arrow::Array>> FixedSizedBinaryDecoder::ToFixedSizeListArray(
+    int32_t start, int32_t length) const {
+  auto fixed_list_type = std::dynamic_pointer_cast<::arrow::FixedSizeListType>(type_);
+  auto value_type = fixed_list_type->value_field();
+  assert(::arrow::is_primitive(value_type->type()->id()));
+  auto plain_decoder = PlainDecoder(infile_, value_type->type());
+  ARROW_RETURN_NOT_OK(plain_decoder.Init());
+
+  auto list_size = fixed_list_type->list_size();
+  plain_decoder.Reset(position_, length_ * list_size);
+  ARROW_ASSIGN_OR_RAISE(auto values, plain_decoder.ToArray(start * list_size, length * list_size));
+  return std::make_shared<::arrow::FixedSizeListArray>(type_, length, values);
+}
+
+::arrow::Result<std::shared_ptr<::arrow::Array>> FixedSizedBinaryDecoder::ToArray(
+    int32_t start, std::optional<int32_t> length) const {
+  if (!length.has_value()) {
+    length = length_ - start;
+  }
+  if (start + length.value() > length_ || start > length_) {
+    return ::arrow::Status::IndexError(
+        fmt::format("{}::ToArray: out of range: start={}, length={}, page_length={}\n",
+                    ToString(),
+                    start,
+                    length.value(),
+                    length_));
+  }
+  if (lance::arrow::is_fixed_size_list(type_)) {
+    return ToFixedSizeListArray(start, length.value());
+  } else if (::arrow::is_fixed_size_binary(type_->id())) {
+    return ToFixedSizeBinaryArray(start, length.value());
+  }
+  return ::arrow::Status::Invalid("Invalid data type: {}\n", type_->ToString());
+}
+
+::arrow::Result<std::shared_ptr<::arrow::Array>> FixedSizedBinaryDecoder::Take(
+    std::shared_ptr<::arrow::Int32Array> indices) const {
+  std::shared_ptr<::arrow::ArrayBuilder> builder;
+  if (lance::arrow::is_fixed_size_list(type_)) {
+    auto list_type = std::dynamic_pointer_cast<::arrow::FixedSizeListType>(type_);
+    ARROW_ASSIGN_OR_RAISE(auto value_builder,
+                          ::lance::arrow::GetArrayBuilder(list_type->value_type()));
+    builder = std::make_shared<::arrow::FixedSizeListBuilder>(pool_, value_builder, type_);
+  } else if (::arrow::is_fixed_size_binary(type_->id())) {
+    builder = std::make_shared<::arrow::FixedSizeBinaryBuilder>(type_);
+  } else {
+    return ::arrow::Status::Invalid("FixedSizeBuilderDecoder::Take: Invalid data type: ", type_);
+  }
+
+  // TODO: Use thread pool
```

Comment:
Should this be done for all decoders? Or just ones where we may end up reading large amounts of data for each scalar?

Reply:
This should be done if we choose to pick "scalar" individually. Usually it works better for the case where scalar is big, i.e., large binary, tensors, images and etc.

Other cases, i.e., for numeric values, it is faster to read a block and then filter them. (i.e., 64KB = 8000 int64 values in one read)